content/numerical/FastFourierTransform.h

```diff
@@ -3,22 +3,31 @@
  * Date: 2019-01-09
  * License: CC0
  * Source: http://neerc.ifmo.ru/trains/toulouse/2017/fft2.pdf (do read, it's excellent)
-   Papers about accuracy: http://www.daemonology.net/papers/fft.pdf, http://www.cs.berkeley.edu/~fateman/papers/fftvsothers.pdf
-   For integers rounding works if $(|a| + |b|)\max(a, b) < \mathtt{\sim} 10^9$, or in theory maybe $10^6$.
+   Accuracy bound from http://www.daemonology.net/papers/fft.pdf
  * Description: fft(a, ...) computes $\hat f(k) = \sum_x a[x] \exp(2\pi i \cdot k x / N)$ for all $k$. Useful for convolution:
    \texttt{conv(a, b) = c}, where $c[x] = \sum a[i]b[x-i]$.
    For convolution of complex numbers or more than two vectors: FFT, multiply
    pointwise, divide by n, reverse(start+1, end), FFT back.
-   For integers, consider using a number-theoretic transform instead, to avoid rounding issues.
+   Let N be $\max(|a|,|b|)$. Is guaranteed safe as long as $N\log_2{N}\max(a)\max(b) < \mathtt{\sim} 10^{16}$ .
+   Consider using number-theoretic transform or FFTMod instead if precision is an issue.
  * Time: O(N \log N), where $N = |A|+|B|-1$ ($\tilde 1s$ for $N=2^{22}$)
  * Status: somewhat tested
  */
 #pragma once
 
 typedef complex<double> C;
+typedef complex<long double> Cd;
 typedef vector<double> vd;
-
-void fft(vector<C> &a, vector<C> &rt, vi& rev, int n) {
+void fft(vector<C> &a, vector<C> &rt) {
+	int n = sz(a), L= 32 - __builtin_clz(n); vi rev(n);
+	rep(i,0,n) rev[i] = (rev[i / 2] | (i & 1) << L) / 2;
+	if (rt.empty()) {
+		rt.assign(n, 1);
```

Comment:
one way you could shorten this a bit is by taking `k` in the below loop by reference

Reply:
I think I don't want to complicate the FFT interface much, as it's possible that we might want to use FFT as a subroutine in functions other than `conv`.

Reply:
Good point. Another option: make `rt` static, and do
```cpp
static vector<C> rt(2, 1);
for (static int k = 2; k < n; k *= 2) {
    rt.resize(n);
    ... // inner loop as before
}
```
Only wins one line, and can use perf testing (maybe the global variable doesn't get LICM'd), but simplifies the API even further.

Reply:
oh.... smart.

I was gonna ask how it worked with changing `n` sizes, and then I realized.